### PR TITLE
JDK-8313244: NM flags handling in configure process

### PR DESCRIPTION
**Pull request by ansteiner**
On AIX we need the -X64 option for NM in the build. The handling is equivalent to the other used tools flags like AR.
This change will replace the quick fix done in https://bugs.openjdk.org/browse/JDK-8312466.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313244](https://bugs.openjdk.org/browse/JDK-8313244): NM flags handling in configure process (**Enhancement** - P4)

### Reviewers
 * [Christoph Langer](https://openjdk.org/census#clanger) (@RealCLanger - **Reviewer**)
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15150/head:pull/15150` \
`$ git checkout pull/15150`

Update a local copy of the PR: \
`$ git checkout pull/15150` \
`$ git pull https://git.openjdk.org/jdk.git pull/15150/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15150`

View PR using the GUI difftool: \
`$ git pr show -t 15150`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15150.diff">https://git.openjdk.org/jdk/pull/15150.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15150#issuecomment-1665312127)

### CODE REVIEW

File: make/autoconf/flags-other.m4 [hunk 1]
```diff
@@ -88,6 +88,18 @@ AC_DEFUN([FLAGS_SETUP_RCFLAGS],
   AC_SUBST(RCFLAGS)
 ])
 
+AC_DEFUN([FLAGS_SETUP_NMFLAGS],
+[
+  # On AIX, we need to set NM flags
+  if test "x$OPENJDK_TARGET_OS" = xaix; then
+    NMFLAGS="-X64"
+  else
```
Comment (RealCLanger) [inline make/autoconf/flags-other.m4:96]: The else case is not needed, as an empty NMFLAGS would be the default. See the handling of RCFLAGS above.

**Review by RealCLanger — COMMENTED** [review]
Looks good overall. I suggest to remove the BUILD_NM variable in this PR as well? We should then wait with a final review until some folks from the Oracle build team are back.

**Review by TheShermanTanker — APPROVED** [review]
Looks good apart from Christoph's comments, I'll preliminarily approve this since my approval won't immediately flag this as ready for integration

File: make/autoconf/flags-other.m4 (hunk 1 above)
Comment (ansteiner) [reply make/autoconf/flags-other.m4]: I removed the needless else.

**Review by MBaesken — COMMENTED** [review]
make/common/NativeCompilation.gmk  uses NM too, is there a reason to avoid the flag there?
flags-other.m4 : the comment 'on AIX ...'  is just stating the obvious;  maybe it should better mention that without -x64 we only process 32bit objects (Processes only 32-bit object files , and this is the default) ; or avoid the comment altogether ?

**Comment by ansteiner** [comment]
> make/common/NativeCompilation.gmk uses NM too, is there a reason to avoid the flag there? flags-other.m4 : the comment 'on AIX ...' is just stating the obvious; maybe it should better mention that without -X64 we only process 32bit objects ('Processes only 32-bit object files' says the manpage , and this is the default on AIX) ; or avoid the comment altogether ?

I didn't see these nm calls. I focused on the AIX specific only. Thanks for mentioning this. When are static libraries and mapfiles used? In addition I'm wondering how the nm call with the -m option in the else condition will work. This option is not known by all nm implementations, as I know.
I add the NMFLAGS to the missing nm calls, also for other target OS and improved the comment.

**Comment by RealCLanger** [comment]
> > make/common/NativeCompilation.gmk uses NM too, is there a reason to avoid the flag there? flags-other.m4 : the comment 'on AIX ...' is just stating the obvious; maybe it should better mention that without -X64 we only process 32bit objects ('Processes only 32-bit object files' says the manpage , and this is the default on AIX) ; or avoid the comment altogether ?
> 
> I didn't see these nm calls. I focused on the AIX specific only. Thanks for mentioning this. When are static libraries and mapfiles used? In addition I'm wondering how the nm call with the -m option in the else condition will work. This option is not known by all nm implementations, as I know. I add the NMFLAGS to the missing nm calls, also for other target OS and improved the comment.

I think the usage of nm in NativeCompilation.gmk happens when building static libs. It is used for e.g. GraalVM builds, as far as I konw. So it is probably not so relevant for AIX at the moment, however, I think the new NM flags macro should be used there, too for completeness, as long as there is no evidence that it would break the NM usage on AIX if it were used with -X64 there.

**Review by erikj79 — COMMENTED** [review]
I think NMFLAGS should be added in NativeCompilation.gmk as well.

**Comment by erikj79** [comment]
> Looks good overall. I suggest to remove the BUILD_NM variable in this PR as well. We should then wait with a final review until some folks from the Oracle build team are back.

What is the motivation for removing BUILD_NM? I believe that would break cross compilation.

**Comment by RealCLanger** [comment]
> > Looks good overall. I suggest to remove the BUILD_NM variable in this PR as well. We should then wait with a final review until some folks from the Oracle build team are back.
> 
> What is the motivation for removing BUILD_NM? I believe that would break cross compilation.

Hm, I guess I didn't/don't really understand how it is used. But alright then 😄

**Comment by ansteiner** [comment]
I agree to add this to the nm calls in NativeCompilation.gmk. Christoph mentioned this is used when static libraries are build. I tried the target "static-libs-image". But this isn't working on AIX, because the ar flags -rcs are not available on AIX. Also the nm calls of this target use the added -X64 already. 
Can somebody tell when this is used to test this too?

**Comment by erikj79** [comment]
> > > Looks good overall. I suggest to remove the BUILD_NM variable in this PR as well. We should then wait with a final review until some folks from the Oracle build team are back.
> > 
> > 
> > What is the motivation for removing BUILD_NM? I believe that would break cross compilation.
> 
> Hm, I guess I didn't/don't really understand how it is used. But alright then 😄

All the `BUILD_` prefixed variables replace the main variables when compiling a "BUILDJDK" during a cross compilation build. This happens in `make/autoconf/buildjdk-spec.gmk.in`. For completeness we should add a BUILD_NMFLAGS, but I think that can wait for now.

**Review by erikj79 — APPROVED** [review]

**Review by RealCLanger — APPROVED** [review]

**Review by TheShermanTanker — APPROVED** [review]

**Review by MBaesken — APPROVED** [review]

**Pull request closed without merging**